PopPUNK/utils.py

```diff
@@ -497,3 +494,162 @@ def isolateNameToLabel(names):
+            Keep sketches of genomes that fail QC separate from main
+            database
+        length_sigma (int)
+            Number of SDs of length distribution beyond which sequences
+            are excluded
+        lower_length (int)
+            Threshold length below which sequences are excluded
+        upper_length (int)
+            Threshold length above which sequences are excluded
+        prop_n (float)
+            Proportion of ambiguous bases above which sequences are excluded
+        upper_n (int)
+            Number of ambiguous bases above which sequences are excluded
+        strand_preserved (bool)
+            Ignore reverse complement k-mers (default = False)
+        threads (int)
+            Number of threads to use in parallelisation
+
+    Returns:
+        genome_length (int)
+            Average length of assemblies
+    """
+    # Genome length needed to calculate prob of random matches
+    if estimated_length is not None:
+        mean_genome_length = estimated_length # assume 2 Mb in the absence of other information
+
+    # open databases
+    db_name = prefix + '/' + os.path.basename(prefix) + '.h5'
+    filtered_db_name = prefix + '/' + 'filtered.' + os.path.basename(prefix) + '.h5'
+    failed_db_name = prefix + '/' + 'failed.' + os.path.basename(prefix) + '.h5'
+    hdf_in = h5py.File(db_name, 'r')
+    # new database file if pruning
+    if qc_filter == 'prune':
+        hdf_out = h5py.File(filtered_db_name, 'w')
+    # retain sketches of failed samples
+    if retain_failures:
+        hdf_fail = h5py.File(failed_db_name, 'w')
+    
+    # process data structures
+    read_grp = hdf_in['sketches']
+    if qc_filter == 'prune':
+        out_grp = hdf_out.create_group('sketches')
+    if retain_failures:
+        fail_grp = hdf_fail.create_group('sketches')
+    seq_length = {}
+    seq_ambiguous = {}
+    seq_excluded = {}
+    removed = []
+    retained = []
+    
+    # iterate through sketches
+    for dataset in read_grp:
+        # test thresholds
+        remove = False
+        seq_length[dataset] = hdf_in['sketches'][dataset].attrs['length']
+        seq_ambiguous[dataset] = hdf_in['sketches'][dataset].attrs['missing_bases']
+        # if no filtering to be undertaken, retain all sequences
+        if qc_filter == 'continue':
+            retained.append(dataset)
+
+    # calculate thresholds
+    with open(prefix + '/' + os.path.basename(prefix) + '_qcreport.txt', 'a+') as qc_file:
+    
+        # get mean length if not using user-supplied estimation
+        if estimated_length is None:
+            genome_lengths = np.fromiter(seq_length.values(), dtype = int)
+            mean_genome_length = np.mean(genome_lengths)
+        
+        # calculate length threshold unless user-supplied
+        if lower_length is None:
+            lower_length = mean_genome_length - length_sigma*np.std(genome_lengths)
+            upper_length = mean_genome_length + length_sigma*np.std(genome_lengths)
```

Comment:
These should probably be outside the `with`

Reply:
Done in e601f73.